File: nodejs/src/client/domains.ts
/**
 * Domains API: list, get.
 * Backend: GET /organizations/domains, GET /organizations/domains/:domain_id.
 */

import type { LoxtepHttpClient } from '../http/client.js';
import type { Domain, DomainsListResponse, DomainsListFilters } from './domains-types.js';

const DOMAINS_BASE = '/organizations/domains';

function buildQueryString(params: Record<string, string | number | undefined>): string {
  const search = new URLSearchParams();
  for (const [k, v] of Object.entries(params)) {
    if (v === undefined || v === null) continue;
    search.set(k, String(v));
  }
  const qs = search.toString();
  return qs ? `?${qs}` : '';
}

export function createDomainsApi(http: LoxtepHttpClient): {
  get: (domain_id: string) => Promise<Domain>;
  list: (filters?: DomainsListFilters) => Promise<DomainsListResponse['data']>;
} {
  return {
    async list(filters?: DomainsListFilters): Promise<DomainsListResponse['data']> {
      const params: Record<string, string | number | undefined> = {
        page: filters?.page ?? 1,
        page_size: filters?.page_size ?? 50,
      };
      if (filters?.organization_id) params.organization_id = filters.organization_id;
      if (filters?.status) params.status = filters.status;
      if (filters?.search) params.search = filters.search;
      const qs = buildQueryString(params);
      const res = await http.get<DomainsListResponse>(`${DOMAINS_BASE}${qs}`);
      return res.data;
    },

    async get(domain_id: string): Promise<Domain> {
      const res = await http.get<{ success: true; data: Domain }>(
        `${DOMAINS_BASE}/${encodeURIComponent(domain_id)}`
      );
      return res.data;
    },
  };
}

File: python/src/loxtep/domains.py
"""
Domains API. list, get.
Backend: organizations microservice /organizations/domains.
"""

from typing import Any, Optional
from urllib.parse import quote

from .http_client import AsyncLoxtepHttpClient, LoxtepHttpClient

BASE = "/organizations/domains"


def _unwrap(res: Any) -> Any:
    return res.get("data", res) if isinstance(res, dict) else res


def _query_string(params: dict[str, Any]) -> str:
    parts = [f"{k}={v}" for k, v in params.items() if v is not None]
    return "?" + "&".join(parts) if parts else ""


class DomainsApi:
    """Sync domains surface."""

    def __init__(self, http: LoxtepHttpClient) -> None:
        self._http = http

    def list(
        self,
        *,
        page: int = 1,
        page_size: int = 50,
        organization_id: Optional[str] = None,
        status: Optional[str] = None,
        search: Optional[str] = None,
    ) -> dict[str, Any]:
        qs = _query_string(
            {
                "page": page,
                "page_size": page_size,
                "organization_id": organization_id,
                "status": status,
                "search": search,
            }
        )
        return _unwrap(self._http.get(f"{BASE}{qs}"))

    def get(self, domain_id: str) -> dict[str, Any]:
        return _unwrap(self._http.get(f"{BASE}/{quote(domain_id)}"))


class AsyncDomainsApi:
    """Async domains surface."""

    def __init__(self, http: AsyncLoxtepHttpClient) -> None:
        self._http = http

    async def list(
        self,
        *,
        page: int = 1,
        page_size: int = 50,
        organization_id: Optional[str] = None,
        status: Optional[str] = None,
        search: Optional[str] = None,
    ) -> dict[str, Any]:
        qs = _query_string(
            {
                "page": page,
                "page_size": page_size,
                "organization_id": organization_id,
                "status": status,
                "search": search,
            }
        )
        return _unwrap(await self._http.get(f"{BASE}{qs}"))

    async def get(self, domain_id: str) -> dict[str, Any]:
        return _unwrap(await self._http.get(f"{BASE}/{quote(domain_id)}"))
